sdk/python/toolchain/uv.go

```diff
@@ -0,0 +1,265 @@
+				virtualenv = filepath.Join(root, defaultVirtualEnv)
+			} else {
+				// We have a pyproject.toml, place the virtualenv next to it.
+				virtualenv = filepath.Join(pyprojectTomlDir, defaultVirtualEnv)
+			}
+		} else {
+			// We have a uv.lock, place the virtualenv next to it.
+			virtualenv = filepath.Join(uvLockDir, defaultVirtualEnv)
+		}
+	}
+	if !filepath.IsAbs(virtualenv) {
+		virtualenv = filepath.Join(root, virtualenv)
+	}
+
+	u := &uv{
+		virtualenvPath: virtualenv,
+		root:           root,
+	}
+
+	// Validate the version
+	cmd := u.uvCommand(context.Background(), "", false, nil, nil, "--version")
+	versionString, err := cmd.Output()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get uv version: %w", err)
+	}
+	version, err := u.uvVersion(string(versionString))
+	if err != nil {
+		return nil, err
+	}
+	logging.V(9).Infof("Python toolchain: using uv version %s", version)
+
+	return u, nil
+}
+
+func (u *uv) InstallDependencies(ctx context.Context, cwd string, useLanguageVersionTools,
+	showOutput bool, infoWriter, errorWriter io.Writer,
+) error {
+	if useLanguageVersionTools {
+		if err := installPython(ctx, cwd, showOutput, infoWriter, errorWriter); err != nil {
+			return err
+		}
+	}
+
+	// Look for a uv.lock file.
+	// If no uv.lock file is found, look for a pyproject.toml file.
+	// If no pyproject.toml file is found, create it, and then look for a
+	// requirements.txt file to install dependencies.
+	if _, err := searchup(cwd, "uv.lock"); err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return fmt.Errorf("error while looking for uv.lock in %s: %w", cwd, err)
+		}
+
+		// No uv.lock found, look for pyproject.toml.
+		if _, err := searchup(cwd, "pyproject.toml"); err != nil {
+			if !errors.Is(err, os.ErrNotExist) {
+				return fmt.Errorf("error while looking for pyproject.toml in %s: %w", cwd, err)
+			}
+
+			// No pyproject.toml found, create one using `uv`
+			initCmd := u.uvCommand(ctx, cwd, showOutput, infoWriter, errorWriter,
+				"init", "--no-readme", "--no-package", "--no-pin-python")
+			if err := initCmd.Run(); err != nil {
+				return fmt.Errorf("error initializing python project: %w", err)
+			}
+
+			// `uv init` creates a `hello.py` file, delete it.
+			contract.IgnoreError(os.Remove(filepath.Join(cwd, "hello.py")))
+
+			// Look for a requirements.txt file and use it to install dependencies
+			requirementsTxtDir, err := searchup(cwd, "requirements.txt")
+			if err == nil {
+				requirementsTxt := filepath.Join(requirementsTxtDir, "requirements.txt")
+				addCmd := u.uvCommand(ctx, cwd, showOutput, infoWriter, errorWriter, "add", "-r", requirementsTxt)
+				if err := addCmd.Run(); err != nil {
+					return fmt.Errorf("error installing dependecies from requirements.txt: %w", err)
+				}
+				// Remove the requirements.txt file, after calling `uv add`, the
+				// dependencies are tracked in pyproject.toml.
+				if err := os.Remove(requirementsTxt); err != nil {
```

Comment:
This seems a little surprising to me.  As a user I don't think I would expect the requirements.txt to be deleted here I think.  At least we should have some output for the user that we did that if we are removing it.

Reply:
Is there any issue with just keeping the `requirements.txt` around? I'm not familiar with Uv, so I'm curious.

Reply:
@tgummerer From the UV side, this doesn't help or hinder; what can happen is that the user is not very familiar with UV and continues to add dependencies in `requirements.txt` instead of adding them to `pyproject.toml` and thinks that Pulumi is not updating the libraries.

Reply:
Maybe the best thing to do is do nothing?

(correct place :smile: )

Reply:
I added a log that tells the user that requirements.txt is deleted and that dependencies are tracked in pyproject.tonl.

The main motivation is that a project generated by `pulumi new` should be neat and tidy, no matter the chosen toolchain, and the templates all use `requirements.txt` as common denominator. Poetry has the same behaviour of deleting requirements.txt.

I'll follow up with adding a log for Poetry as well.